### PR TITLE
src/engine/system.cpp: speed up resolution of unmodified paths on unix

### PR DESCRIPTION
**Pull request by trofi**
Without the change `fheroes2` can start on NixOS for a few minutes.
It happens due to suboptimal interaction of NixOS flat directory layout:

    /nix/store/...-fheroes2-${ver}/share/fheroes2/files/lang
    /nix/store/...-other-package-1/...
    /nix/store/...-other-package-2/...
    ...
    /nix/store/...-other-package-10000/...

and of `bool System::GetCaseInsensitivePath(path, correctedPath)` function.
When `System::GetCaseInsensitivePath("/nix/store/...-fheroes2-${ver}/share/fheroes2/files/lang", out);`
is called `/nix/store` is fully expanded first to the list of file to look
up `...-fheroes2-${ver}` entry in a case insensitive manner.

My system happens to have 400_000 `/nix/store` entries and it takes tens
of seconds to match a file there.

The change adds an optimisitic direct entry lookup to avoid linear
directory scan.

### CODE REVIEW

File: src/engine/system.cpp [hunk 1]
```diff
@@ -401,6 +401,33 @@ bool System::GetCaseInsensitivePath( const std::string & path, std::string & cor
 
         correctedPath.append( delimiter );
 
+        {
```
Comment (ihhub) [inline src/engine/system.cpp:404]: Please remove extra parenthesis here.

File: src/engine/system.cpp [hunk 2]
```diff
@@ -401,6 +401,33 @@ bool System::GetCaseInsensitivePath( const std::string & path, std::string & cor
...
+            // GetCaseInsensitivePath() calls become very expensive on such systems.
+            //
+            // The idea is to try to open current subpath as a directory an avoid
+            // directory traversal altogether. Otherwise fall back to linear
+            // case-insensitive search.
+
+            std::string absSubpath = correctedPath + *subPathIter;
+            DIR * de = opendir( absSubpath.c_str() );
+            if ( de ) {
+                correctedPath += *subPathIter;
```
Comment (ihhub) [inline src/engine/system.cpp:421]: Please use `std::move()` call to avoid extra memory operations.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @trofi , I left 2 comments in this pull request. Could you please take a look at them?

File: src/engine/system.cpp (hunk 2 above)
Comment (trofi) [reply src/engine/system.cpp]: Switched to `std::move( absSubpath )`.

**Review by ihhub — APPROVED** [review]

File: src/engine/system.cpp [hunk 3]
```diff
@@ -401,6 +401,31 @@ bool System::GetCaseInsensitivePath( const std::string & path, std::string & cor
...
+        // Avoid directory traversal and try to probe directory name directly.
+        // Speeds up file lookup when intermediate directories have a lot of
+        // files. Example is NixOS where file layout is:
+        //     /nix/store/...-fheroes2-${ver}/share/fheroes2/files/lang
+        //     /nix/store/...-other-package-1/...
+        //     /nix/store/...-other-package-2/...
+        // It's not uncommon for /nix/store to have tens of thousands files.
+        // GetCaseInsensitivePath() calls become very expensive on such systems.
+        //
+        // The idea is to try to open current subpath as a directory an avoid
```
Comment (oleg-derevenetz) [inline src/engine/system.cpp:413]: I believe there is a typo: `an avoid` -> `and avoid`.

**Review by oleg-derevenetz — CHANGES_REQUESTED** [review]
Hi @trofi Found just one typo in comment, LGTM in general.

File: src/engine/system.cpp (hunk 3 above)
Comment (trofi) [reply src/engine/system.cpp]: Good catch! Updated.

**Review by oleg-derevenetz — APPROVED** [review]

**Comment by ihhub** [comment]
@trofi , thank you so much for the fix!

**Pull request merged**